File: scripts/run-tool.js
#!/usr/bin/env node

// Usage: ./run-tool.js <tool-name> <tool-data>

const path = require("path");
const { readFile, writeFile } = require("fs/promises");
const os = require("os");

// Standard exit codes
const EXIT_SUCCESS = 0;
const EXIT_GENERAL_ERROR = 1;
const EXIT_INVALID_INPUT = 2;
const EXIT_FILE_NOT_FOUND = 3;
const EXIT_PERMISSION_DENIED = 4;
const EXIT_NETWORK_ERROR = 5;
const EXIT_TIMEOUT = 124;
const EXIT_COMMAND_NOT_FOUND = 127;

async function main() {
  try {
    const [toolName, rawData] = parseArgv("run-tool.js");
    const toolData = parseRawData(rawData);

    const rootDir = path.resolve(__dirname, "..");
    await setupEnv(rootDir, toolName);

    const toolPath = path.resolve(rootDir, `tools/${toolName}.js`);
    await run(toolName, toolPath, "run", toolData);
  } catch (err) {
    console.error(err.message);
    process.exit(err.code || EXIT_GENERAL_ERROR);
  }
}

function parseArgv(thisFileName) {
  let toolName = process.argv[1];
  let toolData = null;

  if (toolName.endsWith(thisFileName)) {
    toolName = process.argv[2];
    toolData = process.argv[3];
  } else {
    toolName = path.basename(toolName);
    toolData = process.argv[2];
  }

  if (toolName && toolName.endsWith(".js")) {
    toolName = toolName.slice(0, -3);
  }

  if (!toolData || !toolName) {
    const err = new Error(`Usage: ./run-tool.js <tool-name> <tool-data>`);
    err.code = EXIT_INVALID_INPUT;
    throw err;
  }

  return [toolName, toolData];
}

function parseRawData(data) {
  if (!data) {
    const err = new Error("No JSON data");
    err.code = EXIT_INVALID_INPUT;
    throw err;
  }
  try {
    return JSON.parse(data);
  } catch {
    const err = new Error("Invalid JSON data");
    err.code = EXIT_INVALID_INPUT;
    throw err;
  }
}

async function setupEnv(rootDir, toolName) {
  await loadEnv(path.resolve(rootDir, ".env"));
  process.env["LLM_ROOT_DIR"] = rootDir;
  process.env["LLM_TOOL_NAME"] = toolName;
  process.env["LLM_TOOL_CACHE_DIR"] = path.resolve(rootDir, "cache", toolName);
  if (process.stdout.isTTY) {
    process.env["LLM_OUTPUT_COLOR"] = "1";
  }
}

async function loadEnv(filePath) {
  let lines = [];
  try {
    const data = await readFile(filePath, "utf-8");
    lines = data.split("\n");
  } catch {
    return;
  }

  const envVars = new Map();

  for (const line of lines) {
    if (line.trim().startsWith("#") || line.trim() === "") {
      continue;
    }

    const [key, ...valueParts] = line.split("=");
    const envName = key.trim();

    if (!process.env[envName]) {
      let envValue = valueParts.join("=").trim();
      if ((envValue.startsWith('"') && envValue.endsWith('"')) || (envValue.startsWith("'") && envValue.endsWith("'"))) {
        envValue = envValue.slice(1, -1);
      }
      envVars.set(envName, envValue);
    }
  }

  for (const [envName, envValue] of envVars.entries()) {
    process.env[envName] = envValue;
  }
}

async function run(toolName, toolPath, toolFunc, toolData) {
  if (os.platform() === "win32") {
    toolPath = `file://${toolPath}`;
  }
  
  let mod;
  try {
    mod = await import(toolPath);
  } catch (err) {
    const error = new Error(`Failed to import tool module: ${err.message}`);
    error.code = EXIT_FILE_NOT_FOUND;
    throw error;
  }
  
  if (!mod || !mod[toolFunc]) {
    const error = new Error(`Not module function '${toolFunc}' at '${toolPath}'`);
    error.code = EXIT_GENERAL_ERROR;
    throw error;
  }
  
  let value;
  try {
    value = await mod[toolFunc](toolData);
  } catch (err) {
    // Propagate the tool's exit code if available
    if (err.code) {
      process.exit(err.code);
    }
    throw err;
  }
  
  await returnToLLM(value);
  await dumpResult(toolName);
}

async function returnToLLM(value) {
  if (value === null || value === undefined) {
    return;
  }
  const write = async (value) => {
    if (process.env["LLM_OUTPUT"]) {
      await writeFile(process.env["LLM_OUTPUT"], value);
    } else {
      process.stdout.write(value);
    }
  };
  const type = typeof value;
  if (type === "string" || type === "number" || type === "boolean") {
    await write(value.toString());
  } else if (type === "object") {
    const proto = Object.prototype.toString.call(value);
    if (proto === "[object Object]" || proto === "[object Array]") {
      const valueStr = JSON.stringify(value, null, 2);
      require("assert").deepStrictEqual(value, JSON.parse(valueStr));
      await write(valueStr);
    }
  }
}

async function dumpResult(name) {
  if (!process.env["LLM_DUMP_RESULTS"] || !process.env["LLM_OUTPUT"] || !process.stdout.isTTY) {
    return;
  }
  let showResult = false;
  try {
    if (new RegExp(`\\b(${process.env["LLM_DUMP_RESULTS"]})\\b`).test(name)) {
      showResult = true;
    }
  } catch { }

  if (!showResult) {
    return;
  }

  let data = "";
  try {
    data = await readFile(process.env["LLM_OUTPUT"], "utf-8");
  } catch {
    return;
  }
  process.stdout.write(`\x1b[2m----------------------\n${data}\n----------------------\x1b[0m\n`);
}

main();

File: scripts/run-tool.py
#!/usr/bin/env python3
"""
run-tool.py — Dynamic tool runner for LLM tool integrations.
"""

import os
import re
import json
import sys
import importlib.util
import inspect
from typing import (
    Any,
    Dict,
    Literal,
    Tuple,
    Union,
    get_type_hints,
    get_args,
    get_origin,
)

EXIT_SUCCESS = 0
EXIT_GENERAL_ERROR = 1
EXIT_INVALID_INPUT = 2
EXIT_FILE_NOT_FOUND = 3
EXIT_PERMISSION_DENIED = 4
EXIT_NETWORK_ERROR = 5
EXIT_TIMEOUT = 124
EXIT_COMMAND_NOT_FOUND = 127

class ToolError(Exception):
    def __init__(self, message: str, exit_code: int = EXIT_GENERAL_ERROR):
        super().__init__(message)
        self.exit_code = exit_code

def main() -> None:
    try:
        tool_name, raw_data = parse_argv("run-tool.py")
        tool_data = parse_raw_data(raw_data)
        root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
        setup_env(root_dir, tool_name)
        tool_path = os.path.join(root_dir, "tools", f"{tool_name}.py")
        run(tool_name, tool_path, "run", tool_data)
    except ToolError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(e.exit_code)
    except KeyboardInterrupt:
        print("\n[Aborted by user]", file=sys.stderr)
        sys.exit(130)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(EXIT_GENERAL_ERROR)

def parse_argv(this_file_name: str) -> Tuple[str, str]:
    argv = sys.argv
    if os.path.basename(argv[0]).endswith(this_file_name):
        if len(argv) < 3:
            raise ToolError("Usage: ./run-tool.py <tool-name> <tool-data>", EXIT_INVALID_INPUT)
        tool_name, tool_data = argv[1], argv[2]
    else:
        if len(argv) < 2:
            raise ToolError("Usage: ./run-tool.py <tool-name> <tool-data>", EXIT_INVALID_INPUT)
        tool_name, tool_data = os.path.basename(argv[0]), argv[1]

    if tool_name.endswith(".py"):
        tool_name = tool_name[:-3]
    if not tool_name or not tool_data:
        raise ToolError("Usage: ./run-tool.py <tool-name> <tool-data>", EXIT_INVALID_INPUT)
    return tool_name, tool_data

def parse_raw_data(data: str) -> Dict[str, Any]:
    if not data:
        raise ToolError("No JSON data provided", EXIT_INVALID_INPUT)
    try:
        parsed = json.loads(data)
    except json.JSONDecodeError as exc:
        raise ToolError(f"Invalid JSON data: {exc} (raw data: {data!r})", EXIT_INVALID_INPUT)
    if not isinstance(parsed, dict):
        raise ToolError(f"Expected a JSON object, got {type(parsed).__name__}", EXIT_INVALID_INPUT)
    return parsed

def setup_env(root_dir: str, tool_name: str) -> None:
    load_env(os.path.join(root_dir, ".env"))
    cache_dir = os.path.join(root_dir, "cache", tool_name)
    os.makedirs(cache_dir, exist_ok=True)
    os.environ.update({
        "LLM_ROOT_DIR": root_dir,
        "LLM_TOOL_NAME": tool_name,
        "LLM_TOOL_CACHE_DIR": cache_dir
    })
    if "LLM_OUTPUT" not in os.environ:
        os.environ["LLM_OUTPUT"] = "/dev/stdout" if os.name == "posix" else "CON"

def load_env(file_path: str) -> None:
    try:
        with open(file_path, "r", encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                key, _, val = line.partition("=")
                key, val = key.strip(), val.strip()
                if len(val) >= 2 and val[0] == val[-1] and val[0] in ('"', "'"):
                    val = val[1:-1]
                if key and key not in os.environ:
                    os.environ[key] = val
    except (FileNotFoundError, PermissionError, OSError):
        pass

def run(tool_name: str, tool_path: str, tool_func: str, tool_data: Dict[str, Any]) -> None:
    if not os.path.isfile(tool_path):
        raise ToolError(f"Tool file not found: {tool_path}", EXIT_FILE_NOT_FOUND)
    try:
        spec = importlib.util.spec_from_file_location(os.path.basename(tool_path), tool_path)
        mod = importlib.util.module_from_spec(spec)
        sys.modules[spec.name] = mod
        spec.loader.exec_module(mod)
    except Exception as exc:
        raise ToolError(f"Failed to import tool module: {exc}", EXIT_FILE_NOT_FOUND)

    if not hasattr(mod, tool_func):
        raise ToolError(f"No function '{tool_func}' in '{tool_path}'", EXIT_GENERAL_ERROR)
    
    func = getattr(mod, tool_func)
    tool_data = _coerce_types(func, tool_data)

    try:
        value = func(**tool_data)
    except TypeError as exc:
        tb = sys.exc_info()[2]
        inside_tool = tb and tb.tb_next and (tb.tb_next.tb_frame.f_code is not func.__code__ or tb.tb_next.tb_next is not None)
        if inside_tool: raise
        raise ToolError(f"Tool '{tool_name}' run() called with wrong arguments: {exc}", EXIT_INVALID_INPUT)
    except Exception as exc:
        if hasattr(exc, "exit_code"): sys.exit(exc.exit_code)
        raise

    return_to_llm(value)
    dump_result(tool_name)

def _unwrap_optional(hint: Any) -> Any:
    if get_origin(hint) is Union:
        args = [a for a in get_args(hint) if a is not type(None)]
        if len(args) == 1: return args[0]
    return hint

def _coerce_types(func: Any, data: Dict[str, Any]) -> Dict[str, Any]:
    try:
        hints = get_type_hints(func)
    except Exception:
        try:
            hints = {n: p.annotation for n, p in inspect.signature(func).parameters.items()}
        except Exception:
            return data

    coerced = data.copy()
    for key, value in coerced.items():
        if key not in hints: continue
        target_type = _unwrap_optional(hints[key])
        if get_origin(target_type) is Literal:
            allowed = get_args(target_type)
            if value not in allowed:
                for candidate in allowed:
                    try:
                        if type(candidate)(value) == candidate:
                            coerced[key] = candidate
                            break
                    except: continue
            continue
        if not isinstance(value, (str, int, float, bool)): continue
        if target_type is bool:
            if isinstance(value, str):
                v = value.lower()
                if v in ("true", "1", "yes", "on"): coerced[key] = True
                elif v in ("false", "0", "no", "off"): coerced[key] = False
            elif not isinstance(value, bool): coerced[key] = bool(value)
        elif target_type is int:
            if not isinstance(value, bool):
                try: coerced[key] = int(float(value)) if isinstance(value, str) else int(value)
                except: pass
        elif target_type is float:
            try: coerced[key] = float(value)
            except: pass
        elif target_type is str:
            coerced[key] = str(value)

    for key in ["timeout", "max_time", "connect_timeout"]:
        if key in coerced:
            try: coerced[key] = int(coerced[key])
            except: pass
    return coerced

def return_to_llm(value: Any) -> None:
    if value is None: return
    llm_output = os.environ.get("LLM_OUTPUT", "")
    use_file = llm_output and llm_output not in ("/dev/stdout", "/dev/fd/1", "CON")
    writer = open(llm_output, "w", encoding="utf-8") if use_file else sys.stdout
    try:
        if isinstance(value, (str, int, float, bool)): writer.write(str(value))
        elif isinstance(value, (dict, list)): writer.write(json.dumps(value, indent=2))
    finally:
        if use_file: writer.close()

def dump_result(name: str) -> None:
    if not os.getenv("LLM_DUMP_RESULTS") or not os.getenv("LLM_OUTPUT"): return
    try:
        if not os.isatty(sys.stdout.fileno()): return
    except: return
    try:
        if not re.search(rf"\b({os.environ['LLM_DUMP_RESULTS']})\b", name): return
    except: return
    llm_output = os.environ.get("LLM_OUTPUT", "")
    if llm_output in ("/dev/stdout", "/dev/fd/1", "CON"): return
    try:
        with open(llm_output, "r", encoding="utf-8") as fh:
            print(f"\x1b[2m----------------------\n{fh.read()}\n----------------------\x1b[0m")
    except: pass

if __name__ == "__main__":
    main()
